File: 9-7.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

class Solution {
    static int n = 0;
    static List<List<Integer>> E = new ArrayList<>(102);
    static List<Integer> L = new ArrayList<>();
    static List<Integer> X = new ArrayList<>(Collections.nCopies(102, -1));
    static List<Integer> Y = new ArrayList<>();

    static List<Integer> assign_stone(int k, List<Integer> Y, int[] target) {
        List<Integer> cnt = new ArrayList<>(Collections.nCopies(n, 0));
        for (int i = 0; i < k; i++) {
            cnt.set(Y.get(i), cnt.get(Y.get(i)) + 1);
        }
        List<Integer> answer = new ArrayList<>();
        for (int i = 0; i < k; i++) {
            int u = Y.get(i);
            cnt.set(u, cnt.get(u) - 1);
            if (target[u] - 1 <= cnt.get(u) * 3) {
                answer.add(1);
                target[u] -= 1;
            } else if (target[u] - 2 <= cnt.get(u) * 3) {
                answer.add(2);
                target[u] -= 2;
            } else {
                answer.add(3);
                target[u] -= 3;
            }
        }
        return answer;
    }

    static boolean is_ok(int k, List<Integer> Y, int[] target) {
        List<Integer> cnt = new ArrayList<>(Collections.nCopies(n, 0));
        for (int i = 0; i < k; i++) {
            cnt.set(Y.get(i), cnt.get(Y.get(i)) + 1);
        }
        for (int i = 0; i < n; i++) {
            if (target[i] < cnt.get(i) || cnt.get(i) * 3 < target[i]) {
                return false;
            }
        }
        return true;
    }

    static int dfs(int u) {
        if (E.get(u).isEmpty()) {
            return u;
        }
        int ret = dfs(E.get(u).get(X.get(u)));
        X.set(u, (X.get(u) + 1) % E.get(u).size());
        return ret;
    }

    static int[] solution(int[][] edges, int[] target) {
        n = edges.length + 1;
        for (int i = 0; i < 102; i++) {
            E.add(new ArrayList<>());
        }
        for (int[] edge : edges) {
            E.get(edge[0] - 1).add(edge[1] - 1);
        }
        for (int i = 0; i < n; i++) {
            Collections.sort(E.get(i));
            if (E.get(i).isEmpty()) {
                L.add(i);
            } else {
                X.set(i, 0);
            }
        }
        for (int i = 0; i < 10004; i++) {
            Y.add(dfs(0));
        }
        List<Integer> answer = new ArrayList<>();
        answer.add(-1);
        for (int k = 1; k < 10004; k++) {
            if (is_ok(k, Y, target)) {
                answer = assign_stone(k, Y, target);
                break;
            }
        }
        int[] ans = new int[answer.size()];
        for (int i = 0; i < ans.length; i++) ans[i] = answer.get(i);
        return ans;
    }
}

File: 9-7.cpp
#include <bits/stdc++.h>
using namespace std;

int n = 0;
vector<vector<int>> E(102);
vector<int> L;
vector<int> X(102, -1);
vector<int> Y;

vector<int> assign_stone(int k, vector<int>& Y, vector<int>& target) {
    vector<int> cnt(n, 0);
    for (int i = 0; i < k; i++) {
        cnt[Y[i]] += 1;
    }
    vector<int> answer;
    for (int i = 0; i < k; i++) {
        int u = Y[i];
        cnt[u] -= 1;
        if (target[u] - 1 <= cnt[u] * 3) {
            answer.push_back(1);
            target[u] -= 1;
        }
        else if (target[u] - 2 <= cnt[u] * 3) {
            answer.push_back(2);
            target[u] -= 2;
        }
        else {
            answer.push_back(3);
            target[u] -= 3;
        }
    }
    return answer;
}

bool is_ok(int k, vector<int>& Y, vector<int>& target) {
    vector<int> cnt(n, 0);
    for (int i = 0; i < k; i++) {
        cnt[Y[i]] += 1;
    }
    for (int i = 0; i < n; i++) {
        if (target[i] < cnt[i] || cnt[i] * 3 < target[i]) {
            return false;
        }
    }
    return true;
}

int dfs(int u) {
    if (E[u].empty()) {
        return u;
    }
    int ret = dfs(E[u][X[u]]);
    X[u] = (X[u] + 1) % E[u].size();
    return ret;
}

vector<int> solution(vector<vector<int>> edges, vector<int> target) {
    n = edges.size() + 1;
    for (auto& edge : edges) {
        E[edge[0] - 1].push_back(edge[1] - 1);
    }
    for (int i = 0; i < n; i++) {
        sort(E[i].begin(), E[i].end());
        if (E[i].empty()) {
            L.push_back(i);
        }
        else {
            X[i] = 0;
        }
    }
    for (int i = 0; i < 10004; i++) {
        Y.push_back(dfs(0));
    }
    vector<int> answer = { -1 };
    for (int k = 1; k < 10004; k++) {
        if (is_ok(k, Y, target)) {
            answer = assign_stone(k, Y, target);
            break;
        }
    }
    return answer;
}